modules/weaver-tests/src/smithy4s/weavertests/WeaverHttpRequestTestCase.scala

```diff
@@ -0,0 +1,249 @@
+      .map { expectedBody =>
+        request.bodyText.compile.string.map { responseBody =>
+          assert.eql(expectedBody, responseBody)
+        }
+      }
+      .getOrElse(success.pure[IO])
+
+    val headerAssert =
+      testCase.headers
+        .map { expectedHeaders =>
+          expectedHeaders
+            .map { case (name, value) =>
+              val actual = request.headers.get(CIString(name))
+              actual
+                .map { values =>
+                  val actualValue = values.map(_.value).fold
+                  assert.eql(value, actualValue)
+                }
+                .getOrElse(
+                  failure(s"Header $name was not found in the response.")
+                )
+            }
+            .reduce(_ && _)
+        }
+        .getOrElse(success)
+
+    val expectedUri = baseUri
+      .withPath(
+        Uri.Path.unsafeFromString(testCase.uri)
+      )
+      .withQueryParams(
+        testCase.queryParams.combineAll.map {
+          _.split("=", 2) match {
+            case Array(k, v) =>
+              (
+                k,
+                Uri.decode(
+                  toDecode = v,
+                  charset = StandardCharsets.UTF_8,
+                  plusIsSpace = true
+                )
+              )
+          }
+        }.toMap
+      )
+
+    val uriAssert = assert.eql(expectedUri, request.uri)
+    val methodAssert = assert.eql(
+      testCase.method.toLowerCase(),
+      request.method.name.toLowerCase()
+    )
+
+    List(
+      bodyAssert,
+      headerAssert.pure[IO],
+      uriAssert.pure[IO],
+      methodAssert.pure[IO]
+    ).sequence.map(_.reduce(_ && _))
+  }
+
+  private[weavertests] def makeServerTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO],
+      testCase: HttpRequestTestCase,
+      inputFromDocument: Document.Decoder[I]
+  ): GeneratedTest = {
+    GeneratedTest(
+      name = endpoint.id.toString + "(server|request): " + testCase.id,
+      assertions = {
+        val dummyOutput = DefaultSchemaVisitor(endpoint.output)
+        val input = testCase.params
+          .map { inputFromDocument.decode(_).liftTo[IO] }
+          .getOrElse(IO.unit.asInstanceOf[IO[I]])
+        (deferred[I], input).tupled
+          .flatMap { case (requestDeferred, input) =>
+            type R[I_, E_, O_, SE_, SO_] = IO[O_]
+
+            val fakeImpl: smithy4s.Monadic[Alg, IO] =
+              service.transform[R](
+                new smithy4s.Interpreter[Op, IO] {
+                  def apply[I_, E_, O_, SE_, SO_](
+                      op: Op[I_, E_, O_, SE_, SO_]
+                  ): IO[O_] = {
+                    val (in, endpointInternal) = service.endpoint(op)
+
+                    if (endpointInternal.id == endpoint.id)
+                      requestDeferred.complete(in.asInstanceOf[I]) *>
+                        IO.pure(dummyOutput.asInstanceOf[O_])
+                    else IO.raiseError(new Throwable("Wrong endpoint called"))
```

Comment:
I think we might want to make this go through Document in the future instead of casting, so that it works in Dynamic. It's fine right now